_drafts/algebra-of-types.md

```diff
@@ -0,0 +1,319 @@
+---
+layout: post
+title: "Adventures in Category Theory - Algebra of types"
+categories: learn fp category-theory
+---
+
+[Last time] we have become familiar with the very basics of category theory, and even had a look at some scala code covering isomorphisms. It is time for looking at products, coproducts and the algebra of types.
+
+## Universal construction
+
+Universal construction is common in category theory, it is used to define objects in terms of their relationships to other objects. First, we find a pattern, a shape consisting of objects and morphisms, then look at all its occurrences. There might be many, so we have to find a way to rank those so we can pick the best possible candidate. Let's meet some of these universal constructions!
+
+## Product
+
+The pattern is like this:
+
+![product][product]
+
+So we say an object is good candidate for being a product of two other objects, if there are two morphisms, two _projections_ to them, which picks the first and the second element of the product.
+There might be many such candidates for any two given objects and the best one is which has incoming arrows from all the other ones, so we could define the other projections by composition. (Note, that there might be many such 'best' candidates too, but that means, that they have arrows toward each other as well, so they are isomorphic!)
+
+![product best candidate][product best candidate]
+
+> A product of two objects `a` and `b` is the object `c` equipped with two projections such that for any other object `c'` equipped with two projections there is a unique morphism `m` from `c'` to `c` that factorizes those projections.
+
+On this figure: `p' = p ◦ m` and `q' = q ◦ m`
+
+The canonical product type is the `Tuple2` in scala. It just pairs two types together. We will use the type alias `*` for it, because product is like multiplication, right?
+
+```scala
+object Product {
+  type *[A, B] = (A, B)
+
+  def projectFst[A]: A * _ => A = _._1
+  def projectSnd[B]: _ * B => B = _._2
+}
+
+import Product._
+```
+
+## Coproduct
+
+In category theory, every construction has a dual, an inverse. If we invert the arrows in the definition of a product, we end up with the object `c` equipped with two _injections_ from `a` and `b`. Ranking two possible candidates is also inverted `c` is better than `c'` if there is a unique morphism from `c` to `c'` (so we could define `c'`'s injections by composition)
```

Comment:
What does it mean that `c` is better than `c'` in this context? How is it better? Some additional explanation might help.

Reply:
it means the same as it meant in the discussion of universal constructions, or products, but I will add a few more thoughts to them as well.

Reply:
Indeed, I only gave it a quick glance and manage to skip past that explanation. After a second read, the meaning is clear.